app/controllers/devices_controller.rb

```diff
@@ -0,0 +1,35 @@
+class DevicesController < ApplicationController
+  before_action :authenticate_user!, only: [:create]
```

Comment:
Devices can only be created for authenticated users. Making this the substitute of the [Authenticated Users](https://pusher.com/docs/beams/concepts/authenticated-users) Pusher Beams solution

Reply:
rpush doesn't have a solution for this? I'm curious why we have to implement our own solution

Reply:
In this case it doesn't. I'm happy to share the reasoning below :)

rpush is a _low level_ solution that allows us to send PNs to device tokens. It's an interface for communicating with APNs/FCM/other delivery systems and handles the complexities of this (i.e. delivery failures/retries). It provides models & migrations for facilitating the backend tasks (storing credentials, tokens, notifications, etc) but doesn't go much further than that.

Pusher Beams and 3rd party services are _high level_ solutions and they provide SDKs for backend and mobile to simplify the entire implementation (device tokens and certificates are stored in Pusher's servers). The drawback of this simplification is that we have to do things their way: The mobile SDK for example has a method called `setUserID` in order to restrict notifications to certain users (they call it [Authenticated Users](https://pusher.com/docs/beams/concepts/authenticated-users) flow).

The main problem for us is that we would need to call `setUserID` on their SDK multiple times for different Forems (different user_id for dev.to, forem.dev, etc) and another hiccup in the backends as well when trying provide the flexibility of delivering to many different apps as targets.

In this current situation (implementing rpush) we handle the low level responsibilities of requesting the device tokens (mobile) and storing + delivering the PNs to them (backend + rpush). The fact that we keep track of which token belongs to each user allows us to [send private info in PNs knowing they will only be delivered to them](https://github.com/forem/forem/pull/12419/files#diff-d7514f1a760a5075ca574f567fbda8b2d40c8fc5f3b0513221c93ec67d5f383dR15) and won't be intercepted by others. This is what I meant by:

> Devices can only be created for authenticated users. Making this the substitute of the Authenticated Users Pusher Beams solution